### ISSUE TITLE
Add "Action Items" as a construct to checkup

### ISSUE THREAD

**Issue by carakessler** [checkupjs/checkup#475]
**Summary**

At the bottom of the checkup report in console (also be pulled into a separate section for JSON results) we want to show a list of `Action Items` so a dev running checkup can know what requires immediate attention. Tasks can be highlighted as urgently actionable, phrased in a way that makes it clear to a dev what they are meant to do to fix. Once an `Action Item` is complete, it will no longer show up in the `Action Items` list (unlike tasks, which are purely informational, not prescriptive, and will show up with zero results). For example, if you have zero skipped tests, that result would show up. But if there was an `Action Item` configured to reduce number of skipped tests, and you had zero, it wouldn't show up  as an `Action Item` for you - since you have already done it. 

We want to make it possible for both task authors and those configuring checkup to add items to the `Action Items` list.

**Examples**

Examples of tasks that would benefit from having `Action Items` configured by authors: 
1) Using yarn 
2) Having proper linting set up 
3) Using volta 

^^^^ The theme emerging here is maybe "Project Prerequisites" or things related to the setup/toolchain of your app. 

Examples of why those configuring checkup may want to set up custom `Action Items`: 
1) We are migrating away from pattern _foo_ and it is blocking an important upgrade. We need everyone to work on prioritizing moving off of _foo_
2) There is a cleanup initiative within an app to remove `lint-disables`, we may want to configure a threshold such that prioritizing removing `lint-disables` would come up as an `Action Item` - so if your app has more than _x_ `lint-disables`, it would show up as an `Action Item` 
3) For a task that breaks down the testing pyramid, we may have there be a conditional `Action Item`, such that if you have fewer than _x_ percent of your tests as unit tests, show an `Action Item` to prioritize refactoring your test suite

^^^^^ theme emerging here is problematic patterns (potentially blocking other work) and overall health of an area of the app.

**Requirements** 
1) `Action Items` are conditional (by default in a boolean manner - i.e number of problems > 0, but can be configured more explicitly as defined by the task)
2) `Action Items` can be turned on by default in a task by the task author 
3) Tasks that are turned on as `Action Items` by default can be overridden by config
4) Tasks that aren't turned on as `Action Items` by default can still have an `ActionItemMessage`, in the event that they do become `Action Items`. That message is also overridable by config
5) Tasks that aren't `Action Items` by default can become `Action Items` by turning them on as such in the config. 
6) There is a max # of `Action Items` (maybe 5?). If more than that are turned on by default given the plugins a user has enabled, 5 are chosen at random (we should also supply a warning message saying that we dropped `Action Items` because there are more than 5 enabled by default). Config will validate that no more than 5 can be passed in. 

**Changes to be made**
```
export interface TaskMetaData {
  taskName: TaskName;
  friendlyTaskName: TaskName;
  taskClassification: TaskClassification;
}

```
will become 
```
export interface TaskMetaData {
  taskName: TaskName;
  friendlyTaskName: TaskName;
  taskClassification: TaskClassification;
  actionItemData?: ActionItemData;
}

export interface ActionItemData {
  isActionItem: boolean;
  actionItemMessage: string;
  actionItemConditions?: ActionItemCondition[];
}

export interface ActionItemCondition { 
  key: string;
  condition: Function;
  value?: number; 
}
```
Task authors can choose to expose any number of `actionItemConditions` to be configured to turn on/off. 

Examples: 
For the `eslint-disable` task, an `ActionItemCondition` could be:
``` 
let condition: ActionItemCondition = {
	key: 'numEslintDisables';
    condition: (numEslintDisables, value = 0) => numEslintDisables > value;
    value: 20; 
}
```
^^ so if there are more than 20 'numEslintDisables', it would show up as an action item. 

The config to make `eslint-disables` an `Action Item` with its default message would look like this: 
```
{
	actionItems: { 
		'eslint-disables': true,
	}

}
```

The config to make `eslint-disables` an `Action Item` with a custom error message would look like this:  
```
{
	actionItems: { 
		'eslint-disables': {
			message: 'Stop disabling eslint!!!!!!!!!'
		}
	}

}
```
^^^ if you supply a custom message, it is turned on as an `Action Item` by default. If this is confusing, we can also add an `enabled` property, but i dont see why someone would supply a custom message without turning it on.......

The config to _conditionally_ make `eslint-disables` an `Action Item` with a custom error message would look like this:  
```
{
	actionItems: { 
		'eslint-disables': {
			message: 'Stop disabling eslint!!!!!!!!!',
			condition: 'numEslintDisables',
			value: 20
		}
	}

}
```

This change would require verifying config upfront to ensure that the conditions being used in the config are in fact exposed by a given task.

I am open to the idea that adding this level of flexibility may be more complexity that it is worth, and would only be used by power users. I think it is quite powerful and can see many uses for it in my app, but I can ack that it is complicated. We can potentially reduce the scope here by just making action items toggleable on/off in a boolean manner, and maybe adding the custom message as well, without the conditional piece (basically removing `ActionItemCondition` from `ActionItemData`). 

Looking forward to feedback :)

**Comment by carakessler** [comment]
Since we are removing the concept of `Priority` from `TaskClassification` -> I would actually recommend nesting `ActionItem` under that. 

So currently: 

```
export interface TaskMetaData {
  taskName: TaskName;
  friendlyTaskName: TaskName;
  taskClassification: TaskClassification;
}

export type TaskClassification = {
  category: Category;
  priority: Priority;
};
```
will become
```
export interface TaskMetaData {
  taskName: TaskName;
  friendlyTaskName: TaskName;
  taskClassification: TaskClassification;
}

export type TaskClassification = {
  category: string;
  actionItemData?: ActionItemData;
};

```

**Comment by scalvert** [comment]
This is a good start to this concept. I would, perhaps, take a different approach - one that is slightly more compositional based. 

I'd opt to define an interface that would be complimentary to the `Task` interface (specific details TBD):

```ts
interface ActionableTask {
  readonly isActionable: boolean;
  readonly actions: [TaskName, string[]];
}
```

This would allow task authors to decorate their tasks with this interface. We could then introspect the task when running each task in the `TaskList`, and determine if it's actionable, and gather the actions there, similar to how we gather errors. This would allow us to build up the list of actions to be output after all tasks are run. 

Of course, this proposal omits a number of details:

- What sort of data do we capture in actions?
- What does a single action's data structure look like?

But it does allow for a lot of flexibility in how one gathers and populates actions.

**Comment by carakessler** [comment]
We are aligned on extending the `Task` interface to create an `ActionableTask`. 

```
export interface Action {
  key: string;
  condition: Function;
  value?: number;
  message: string;
}

export interface ActionableTask extends Task {
  readonly isActionable: boolean;
  readonly actions: [TaskName, Action[]];
}

**Comment by carakessler** [comment]
Another thought - evaluating the actions has much more to do with the result than the task itself - instead of defining the actions in the Task, and then having to pass them to the TaskResult to be evaluated, maybe this whole thing should take place in the result. So instead of `ActionableTask extends Task`, it would be `ActionableTaskResult extends TaskResult` with the same structures described above

**Comment by scalvert** [comment]
So are you proposing having actions as part of the `TaskResult` itself? We could do this, but you'd have to extract the actions in order to output them. Maybe that would be fine though, as you may want to serialize them to JSON for the output?

**Comment by carakessler** [comment]
ya in that case, we would gather the actions in the `getReporter` function, as they would be a part of the taskResults
